doc/lexer-modes/src/lexer.rs

```diff
@@ -0,0 +1,205 @@
+use std::cell::RefCell;
+use std::fmt;
+use std::rc::Rc;
+
+/// The lexer's mode.
+///
+/// This is shared between the lexer and the parser.
+#[derive(Debug)]
+pub struct LexerMode {
+    /// If `Some`, the next N characters should be returned as a
+    /// `Literal` token.
+    pub literal: Option<usize>,
+}
+
+impl Default for LexerMode {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl LexerMode {
+    pub fn new() -> Self {
+        Self { literal: None }
+    }
+}
+
+#[derive(PartialEq, Eq, Debug)]
+pub enum LexicalError {
+    LengthOverflow(String),
+    TruncatedInput(String),
+}
+
+impl fmt::Display for LexicalError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?}", self)
+    }
+}
+
+pub type Spanned<Token, Loc, LexicalError> = Result<(Loc, Token, Loc), LexicalError>;
+
+// The type of the parser's input.
+//
+// The parser iterators over tuples consisting of the token's starting
+// position, the token itself, and the token's ending position.
+pub(crate) type LexerItem<Token, Loc, LexicalError> = Spanned<Token, Loc, LexicalError>;
+
+#[derive(Debug, Clone, PartialEq)]
+#[allow(non_camel_case_types)]
```

Comment:
Nitpick: beside history, I don't see a good reason why token should be spelled in upper case.

Reply:
In Sequoia, we choose this convention to make the distinction between tokens and productions in the grammar clearer.

Reply:
To be clear: I'm not saying that you or anyone else should adopt this convention. I'm explaining that I have a good reason to use this convention.

Reply:
But you need to write a mapping in the `grammar.lalrpop`, don't you? So you could do:

```
        LPAREN => lexer::Token::LParen,
```

And keep the Rust-side camel case. Anyway, as I said, it's a nitpick - I'm fine keeping it like that, and I surely don't want to block this PR on that :slightly_smiling_face:

Reply:
I could, but I don't 🙂.  This isn't my project. I'm happy for you to modify the code to suit your tastes.

Reply:
Sure, it's just that

> In Sequoia, we choose this convention to make the distinction between tokens and productions in the grammar clearer.

My point was that you can still keep the distinction between tokens and productions in the grammar clear (as in the example above), while using camel case for the Rust enum. As often, I suspect the actual answer is mostly that it's a matter of taste, and that's ok :slightly_smiling_face: 

Honestly this is quite unimportant, so let's merge like this. Thanks again for the contribution!